### ISSUE TITLE
Bug with block optimization

### ISSUE THREAD

**Issue by mrexodia** [icicle-emu/icicle-emu#64]
While working with icicle (python bindings) we found a strange bug that is only triggered when setting `config.optimize_block = true`. The setup is as follows:
- Execute the program one step at a time
- Modify dead registers

Disassembly listing:

```assembly
.text:0000000140001A73 loc_140001A73:
.text:0000000140001A73                 shr     r10d, 7
.text:0000000140001A77                 and     r10d, 1Fh
.text:0000000140001A7B                 jz      short loc_140001A85
.text:0000000140001A7D                 mov     eax, r10d
.text:0000000140001A80                 mov     [rsi+rax*8+8], rdx
.text:0000000140001A85 loc_140001A85:
.text:0000000140001A85                 add     r9, 4
.text:0000000140001A89 loc_140001A89:
.text:0000000140001A89                 mov     [rsi], r9
.text:0000000140001A8C                 mov     rcx, r9
.text:0000000140001A8F loc_140001A8F:
.text:0000000140001A8F                 mov     r10d, [rcx]    ; set r9 before this instruction (or any of the next 3)
.text:0000000140001A92                 mov     eax, r10d
.text:0000000140001A95                 not     eax
.text:0000000140001A97                 mov     r9, rcx    ; this instruction is supposed to overwrite r9 with rcx
.text:0000000140001A9A                 test    al, 3           ; bug happens here!
```

Reproduction code:

https://github.com/icicle-emu/icicle-python/blob/2f06c21de178b16cd339b920a7fac1377576ee1a/src/main.rs#L108-L159

The output looks like this:

```
[0] RIP: 0x140001a73, RCX: 0x0, R9: 0x71000
[1] RIP: 0x140001a77, RCX: 0x0, R9: 0x71000
[2] RIP: 0x140001a7b, RCX: 0x0, R9: 0x71000
[3] RIP: 0x140001a85, RCX: 0x0, R9: 0x71000
[4] RIP: 0x140001a89, RCX: 0x0, R9: 0x71004
[5] RIP: 0x140001a8c, RCX: 0x0, R9: 0x71004
[6] RIP: 0x140001a8f, RCX: 0x71004, R9: 0x71004
[7] RIP: 0x140001a92, RCX: 0x71004, R9: 0x13370900
[8] RIP: 0x140001a95, RCX: 0x71004, R9: 0x13370900
[9] RIP: 0x140001a97, RCX: 0x71004, R9: 0x13370900
[10] RIP: 0x140001a9a, RCX: 0x71004, R9: 0x13370900
[BUG] expected rcx(0x71004) == r9(0x13370900)
```

I confirmed that commenting out the call to `icicle_cpu::lifter::optimize::const_prop` 'fixes' the issue, which means that the issue is probably at least related to the block constant propagation. The `reg_write` function forwards to `icicle_cpu::cpu::write_reg` with the appropriate `VarNode`.

**Comment by mchesser** [comment]
Yeah, an assumption that the block optimizer makes is that there will be no external modifications to registers within individual blocks. This issue can be reproduced with something as simple as:

```
[4c 89 c9] "MOV RCX, R9"
[49 89 c9] "MOV R9, RCX"
```

Which optimizes to the following Pcode:

```
<L0> (entry=0x0):
	instruction(0x0)
	RCX = R9
	instruction(0x3)
```

Without this assumption, it is essentially impossible to perform any multi-instruction optimizations.

---

That said there are a couple of ways that this could be made better:

1. Disable block-level optimizations by default (or remove them completely) -- in general they don't actually help that much for performance and after 802ab0c, most of the optimizations should already be done by Cranelift, which affects full-block execution anyway.
2. Only apply optimizations to the JIT-blocks, meaning the interpreter (which is what is used for single stepping) executes the unoptimized pcode. The emulator checks some preconditions to ensure that the block will either be executed to completion (or to some known exit points) before entering the JIT.
3. Warn when "reg_write" is performed inside of a block when block-level optimizations are enabled.
4. Automatically de-optimize the block if a manual reg-write is performed in it.
5. Add more control over the optimization process (e.g. const-prop flags, but not GPRs).

I'm thinking that (1.) is probably the best solution the near term. Currently the main reason for the block-level optimizations is to simplify some static analysis that occurs later in my other work (fuzzing) and make the pcode more "readable".

However, I've wanted to move towards (2.) for some time, to allow for more aggressive optimizations in certain cases.

**Comment by mrexodia** [comment]
Yeah that totally makes sense, thanks for the quick reply! I also disabled block level optimizations per default in https://github.com/icicle-emu/icicle-python/commit/e749fe1571c1164ddf406fd76c34e6ebcce2a50c. Probably (3) is something I can do with a python exception to avoid footguns, so that's also a good point! Option (4) would also be cool, but probably too much work (especially if the performance gains are not worth it).

> Only apply optimizations to the JIT-blocks, meaning the interpreter (which is what is used for single stepping)

This is also interesting because `jit=False` causes a super weird bug with `ret` not setting RIP to the return address, but instead setting it to the start of the basic block. I'm still trying to get a minimal reproduction case, but it's super tricky so it might take a while. This only happens with single stepping and setting `jit=True` does not exhibit this behavior.

**Comment by mchesser** [comment]
I was able to reproduce something that seems similar to your other issue, with this:

```rust
let mut vm = crate::build(&Config {
    triple: "riscv64-none".parse().unwrap(),
    enable_jit: false,
    enable_shadow_stack: false,
    ..Config::default()
})
.unwrap();
vm.cpu.mem.map_memory_len(0x1000, 0x100, Mapping { perm: perm::READ | perm::EXEC, value: 0 });

// Map some code that jumps to the target location.
static CODE: &[u8] = &[
    0xb7, 0x10, 0x00, 0x00, // lui      ra,0x1
    0x9b, 0x80, 0x00, 0x01, // addiw    ra,ra,0x10
    0x67, 0x80, 0x00, 0x00, // ret
];
vm.cpu.mem.write_bytes(0x1000, CODE, perm::NONE).unwrap();

// Ensure decode will fail at the jump destination.
vm.cpu.mem.update_perm(0x1010, 4, perm::NONE).unwrap();

vm.cpu.write_pc(0x1000);
vm.step(3);
assert_eq!(vm.cpu.read_pc(), 0x1010);
// At this point there is a pending 'ExecViolation' exception for the next instruction, however
// this is suppressed because we reached the instruction limit.

// If we try to step from this point we should then see the missing exception. However,
assert_eq!(vm.step(1), VmExit::UnhandledException((ExceptionCode::ExecViolation, 0x1010))); // This assertion fails
// And the PC should remain the same.
assert_eq!(vm.cpu.read_pc(), 0x1010); // This assertion will also fails.
```

The issue is related to the fact that emulator attempts to decode the destination of a branch when it is reached rather than when the first instruction is executed (I think the original reason, was related to propagating SLEIGH context from the parent block to child blocks). 

If the destination is a valid this doesn't cause any issues, however if it there is an exception generated during decoding it needs to be handled. During normal execution we can assume that the next instruction will be executed so the emulator immediately returns the error, however during single stepping the instruction hasn't been executed, so returning the exception would be wrong and needs to wait until the next step. To account for this, the emulator checks the icount limit at the point the exception would be generated, and returns an `InstructionLimit` exit instead.

However, this case leaves the `block_id` of the block that was just executed alone, and the offset within the block is only updated if execution stops in the middle of the block, causing the interpreter to start re-executing the block from the beginning.

The reason it works when the JIT is enabled is partially just by chance. For performance reasons on the fast path, the JIT uses the current virtual address (which is re-read from the current PC) to lookup the next function to execute (ignoring the block id).

**Comment by mchesser** [comment]
The broken pc issue should be fixed by: 876009d

**Comment by mrexodia** [comment]
A friend found a related(?) issue where calling `cpu.write_reg` with the `RIP` (PC) register appears to not count instructions correctly and steps too far: https://github.com/icicle-emu/icicle-python/pull/11/commits/d93a4192f0d0ae5e81d01152b51db2b4ed4ce526

This one I can work around in the bindings by checking if the found register offset is the pc_offset and then calling `cpu.write_pc` instead, but it's definitely a bit of a footgun 😅
